# Random Forest Model

In [139]:
# Update sklearn to prevent version mismatches
!pip install sklearn --upgrade

Requirement already up-to-date: sklearn in /Users/bunmiolakanmi/opt/anaconda3/lib/python3.8/site-packages (0.0)


In [140]:
# install joblib. This will be used to save your model. 
# Restart your kernel after installing 
!pip install joblib

In [141]:
import pandas as pd

# Read the CSV and Perform Basic Data Cleaning

In [142]:
df = pd.read_csv("exoplanet_data.csv")
# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')
# Drop the null rows
df = df.dropna() 
df.head()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 41 columns]

In [143]:
df.columns

Index(['koi_disposition', 'koi_fpflag_nt', 'koi_fpflag_ss', 'koi_fpflag_co',
       'koi_fpflag_ec', 'koi_period', 'koi_period_err1', 'koi_period_err2',
       'koi_time0bk', 'koi_time0bk_err1', 'koi_time0bk_err2', 'koi_impact',
       'koi_impact_err1', 'koi_impact_err2', 'koi_duration',
       'koi_duration_err1', 'koi_duration_err2', 'koi_depth', 'koi_depth_err1',
       'koi_depth_err2', 'koi_prad', 'koi_prad_err1', 'koi_prad_err2',
       'koi_teq', 'koi_insol', 'koi_insol_err1', 'koi_insol_err2',
       'koi_model_snr', 'koi_tce_plnt_num', 'koi_steff', 'koi_steff_err1',
       'koi_steff_err2', 'koi_slogg', 'koi_slogg_err1', 'koi_slogg_err2',
       'koi_srad', 'koi_srad_err1', 'koi_srad_err2', 'ra', 'dec',
       'koi_kepmag'],
      dtype='object')

In [144]:
# import os
# os.getcwd()

# Select your features (columns)

In [68]:
# from sklearn.feature_selection import SelectFromModel
# sel = SelectFromModel(RandomForestClassifier(n_estimators = 100))
# sel.fit(X_train, y_train)

In [69]:
# from sklearn.feature_selection import VarianceThreshold
# X= df.drop(columns=['koi_disposition'])
# X
# sel = VarianceThreshold(threshold=(.8 * (1 - .8)))
# x=sel.fit_transform(X)
# x

In [145]:
# Select features (columns). This will also be used as your x values.
selected_X = df[['koi_period',  'koi_time0bk', 'koi_depth', 'koi_prad', 'koi_teq', 'koi_insol', 'koi_model_snr', 'koi_tce_plnt_num', 'koi_steff', 'koi_slogg', 'koi_srad', 'ra', 'dec', 'koi_kepmag']]
selected_X.head()

koi_period  koi_time0bk  koi_depth  koi_prad  koi_teq  koi_insol  \
0   54.418383   162.513840      874.8      2.83      443       9.11   
1   19.899140   175.850252    10829.0     14.60      638      39.30   
2    1.736952   170.307565     8079.2     33.46     1395     891.96   
3    2.525592   171.595550      603.3      2.75     1406     926.16   
4    4.134435   172.979370      686.0      2.77     1160     427.65   

   koi_model_snr  koi_tce_plnt_num  koi_steff  koi_slogg  koi_srad         ra  \
0           25.8                 2       5455      4.467     0.927  291.93423   
1           76.3                 1       5853      4.544     0.868  297.00482   
2          505.6                 1       5805      4.564     0.791  285.53461   
3           40.9                 1       6031      4.438     1.046  288.75488   
4           40.2                 2       6046      4.486     0.972  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714

# Create a Train Test Split

Use `koi_disposition` for the y values

In [129]:
# X= df.drop(columns=['koi_disposition'])
# X

In [146]:
X = selected_X
X

koi_period  koi_time0bk  koi_depth  koi_prad  koi_teq  koi_insol  \
0      54.418383   162.513840      874.8      2.83      443       9.11   
1      19.899140   175.850252    10829.0     14.60      638      39.30   
2       1.736952   170.307565     8079.2     33.46     1395     891.96   
3       2.525592   171.595550      603.3      2.75     1406     926.16   
4       4.134435   172.979370      686.0      2.77     1160     427.65   
...          ...          ...        ...       ...      ...        ...   
6986    8.589871   132.016100       87.7      1.11      929     176.40   
6987    0.527699   131.705093     1579.2     29.35     2088    4500.53   
6988    1.739849   133.001270       48.5      0.72     1608    1585.81   
6989    0.681402   132.181750      103.6      1.07     2218    5713.41   
6990    4.856035   135.993300       76.7      1.05     1266     607.42   

      koi_model_snr  koi_tce_plnt_num  koi_steff  koi_slogg  koi_srad  \
0              25.8                 2       5455      4.467     0.927   
1              76.3                 1       5853      4.544     0.868   
2             505.6                 1       5805      4.564     0.791   
3              40.9                 1       6031      4.438     1.046   
4              40.2                 2       6046      4.486     0.972   
...             ...               ...        ...        ...       ...   
6986            8.4                 1       5638      4.296     1.088   
6987          453.3                 1       5638      4.529     0.903   
6988           10.6                 1       6119      4.444     1.031   
6989           12.3                 1       6173      4.447     1.041   
6990            8.2                 1       6469      4.385     1.193   

             ra        dec  koi_kepmag  
0     291.93423  48.141651      15.347  
1     297.00482  48.134129      15.436  
2     285.53461  48.285210      15.597  
3     288.75488  48.226200      15.509  
4     296.28613  48.224670      15.714  
...         ...        ...         ...  
6986  298.74921  46.973351      14.478  
6987  297.18875  47.093819      14.082  
6988  286.50937  47.163219      14.757  
6989  294.16489  47.176281      15.385  
6990  297.00977  47.121021      14.826  

[6991 rows x 14 columns]

In [147]:
y = df["koi_disposition"]
target_names = ['CONFIRMED', 'FALSE POSITIVE', 'CANDIDATE']

## Pre-processing

In [148]:
# Split data to train and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [153]:
X_train

koi_period  koi_time0bk  koi_depth  koi_prad  koi_teq  koi_insol  \
6122    6.768901   133.077240      123.1      1.24     1017     253.30   
6370    0.733726   132.020050      114.6      0.86     1867    2891.64   
2879    7.652707   134.460380      641.1      3.21      989     226.81   
107     7.953547   174.662240      875.4      2.25      696      55.37   
29      4.959319   172.258529     9802.0     12.21     1103     349.40   
...          ...          ...        ...       ...      ...        ...   
3772    8.268081   135.056330    71479.0     50.31      863     130.83   
5191   11.161938   133.553800      124.4      1.97     1093     337.23   
5226    6.150251   134.422825     2128.7    100.03     2251    6066.49   
5390    3.343285   134.845100      166.8     64.00     2094    4535.97   
860     2.271907   134.798520      149.0      1.12     1318     714.27   

      koi_model_snr  koi_tce_plnt_num  koi_steff  koi_slogg  koi_srad  \
6122           10.8                 1       5737      4.327     1.125   
6370           13.8                 1       5855      4.578     0.797   
2879          254.3                 1       6328      4.481     0.963   
107            38.4                 1       4768      4.536     0.779   
29            696.5                 1       5712      4.359     1.082   
...             ...               ...        ...        ...       ...   
3772         1320.5                 1       5749      4.502     0.922   
5191           13.4                 3       6200      4.072     1.640   
5226          471.0                 1       8914      3.896     2.867   
5390           79.5                 2       6541      3.773     2.652   
860            36.8                 1       5788      4.506     0.856   

             ra        dec  koi_kepmag  
6122  294.40472  39.351681      14.725  
6370  284.50391  42.463860      15.770  
2879  295.50211  38.983540      13.099  
107   291.15878  40.750271      15.660  
29    292.16705  48.727589      15.263  
...         ...        ...         ...  
3772  292.53125  46.728699      15.768  
5191  295.21268  49.562180      13.374  
5226  297.18176  45.988441      10.622  
5390  296.86258  41.147419      13.276  
860   294.73419  40.861420      14.459  

[5243 rows x 14 columns]

In [154]:
# Scale your data
from sklearn.preprocessing import StandardScaler
X_scaler = StandardScaler().fit(X_train)

In [155]:
# Transform the training and testing data using the X_scaler and y_scaler models
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Train a Random Forest Model

In [158]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()
rf = rf.fit(X_train_scaled, y_train)
rf.score(X_test_scaled, y_test)

0.7334096109839817

In [165]:
print(f"Training Data Score: {rf.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {rf.score(X_test_scaled, y_test)}")

Training Data Score: 1.0
Testing Data Score: 0.7334096109839817


In [160]:
# Get important variables in the RF
feature_names = X_train.columns
sorted(zip(rf.feature_importances_, feature_names), reverse=True)

[(0.14565763022306397, 'koi_model_snr'),
 (0.1415350403722495, 'koi_prad'),
 (0.09581455802000041, 'koi_depth'),
 (0.08468950494437104, 'koi_period'),
 (0.07143541457633178, 'koi_teq'),
 (0.06838259617882514, 'koi_insol'),
 (0.0681081008424058, 'koi_time0bk'),
 (0.05852001688511494, 'ra'),
 (0.054591626855783344, 'koi_steff'),
 (0.050040816158218385, 'dec'),
 (0.0484681171768035, 'koi_kepmag'),
 (0.04737586967300307, 'koi_srad'),
 (0.04508152498316334, 'koi_slogg'),
 (0.020299183110665697, 'koi_tce_plnt_num')]

In [161]:
# Calculate classification report - RF
from sklearn.metrics import classification_report
predictions = rf.predict(X_test_scaled)
print(classification_report(y_test, predictions,
                            target_names=target_names))

                precision    recall  f1-score   support

     CONFIRMED       0.57      0.46      0.51       411
FALSE POSITIVE       0.76      0.78      0.77       484
     CANDIDATE       0.78      0.84      0.81       853

      accuracy                           0.73      1748
     macro avg       0.70      0.69      0.70      1748
  weighted avg       0.72      0.73      0.73      1748



In [166]:
# # Perform feature selection and fit a random forest model
# from sklearn.feature_selection import SelectFromModel
# from sklearn.ensemble import RandomForestClassifier
# from sklearn.pipeline import Pipeline
# from sklearn.svm import LinearSVC
# from sklearn.impute import SimpleImputer
# clf = Pipeline([
#   ('imputer', SimpleImputer(strategy='median')),
#   ('feature_selection', SelectFromModel(LinearSVC(penalty="l2"))),
#   ('classification', RandomForestClassifier())
# ])
# clf.fit(X_train_scaled, y_train)

In [167]:
# comp=pd.DataFrame()
# comp['y_test']=y_test
# comp['y_pred']=y_pred
# comp

# Hyperparameter Tuning

Use `GridSearchCV` to tune the model's parameters

In [168]:
param_grid = { 
    'n_estimators': [200, 500],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth' : [4,5,6,7,8],
    'criterion' :['gini', 'entropy']
}

In [170]:
from sklearn.model_selection import GridSearchCV
rf_cv = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5)
rf_cv.fit(X_train_scaled, y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [4, 5, 6, 7, 8],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'n_estimators': [200, 500]})

In [173]:
print(rf_cv.best_params_)
print(rf_cv.best_score_)

{'criterion': 'entropy', 'max_depth': 8, 'max_features': 'auto', 'n_estimators': 500}
0.73164154883968


In [174]:
# Make predictions with the hypertuned model
predictions2= rf_cv.predict(X_test_scaled)

In [175]:
print('Test Acc: %.3f' % rf_cv.score(X_test_scaled, y_test))

Test Acc: 0.724


In [176]:
# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions2,
                            target_names))

                precision    recall  f1-score   support

     CONFIRMED       0.68      0.84      0.75       484
FALSE POSITIVE       0.79      0.81      0.80       853
     CANDIDATE       0.60      0.41      0.49       411

      accuracy                           0.72      1748
     macro avg       0.69      0.69      0.68      1748
  weighted avg       0.72      0.72      0.71      1748



/Users/bunmiolakanmi/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:67: FutureWarning: Pass labels=['CONFIRMED', 'FALSE POSITIVE', 'CANDIDATE'] as keyword args. From version 0.25 passing these as positional arguments will result in an error
  warnings.warn("Pass {} as keyword args. From version 0.25 "


# Save the Model

In [179]:
# save your model by updating "your_name" with your name
# and "your_model" with your model variable
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = 'RandomForest.sav'
joblib.dump(rf, filename)

['RandomForest.sav']

In [178]:
# import os
# os.getcwd()

#### Inference(s)
* The random forest model performance after tuning (0.724) is lower than its performance before tuning (0.733)
* The random forest model is the best performing classifier among all other classifiers built (knn, svc, decision tree and logisticc regression).